### PR TITLE
Implement secure gopher protocol.

### PR DESCRIPTION
**Pull request by parazyd**
This commit introduces a "gophers" handler inside the gopher protocol if
USE_SSL is defined. This protocol is no different than the usual gopher
prococol, with the added TLS encapsulation upon connecting. The protocol
has been adopted in the gopher community, and many people have enabled
TLS in their gopher daemons like geomyidae(8), and clients, like clic(1)
and hurl(1).

I have not implemented test units for this protocol because my knowledge
of Perl is sub-par. However, for someone more knowledgeable it might be
fairly trivial, because the same test that tests the plain gopher
protocol can be used for "gophers" just by adding a TLS listener.

Signed-off-by: parazyd <parazyd@dyne.org>

### CODE REVIEW

**Comment by parazyd** [comment]
As for manual testing, it can be tested with servers that support TLS:

```
curl gophers://parazyd.org
```

**Comment by emilengler** [comment]
Sorry, I disagree. So far there is no specification for it.
In March this year I brought up a discussion at the Gopher-ML about a specification (https://lists.debian.org/gopher-project/2020/03/msg00005.html)
As long as there isn't anything to agree on this shouldn't be merged in my opinion.

**Comment by emilengler** [comment]
Ping @sebdeckers, maybe interesting for you?

File: lib/gopher.c [hunk 1]
```diff
@@ -75,6 +80,49 @@ const struct Curl_handler Curl_handler_gopher = {
...
+static CURLcode gopher_connect(struct connectdata *conn, bool *done)
+{
+  (void)conn;
+  (void)done;
+  return CURLE_OK;
+}
+
+static CURLcode gopher_connecting(struct connectdata *conn, bool *done)
+{
+  CURLcode result;
```
Comment (emilengler) [inline lib/gopher.c:115]: Why not use shotgun initialization here?

Comment (parazyd) [reply lib/gopher.c:115]: I'm new to the curl codebase, and there's bound to be mistakes.

I initially thought that TLS support could simply be added by defining the struct, and only setting the `PROTOPT_SSL` flag, but it seems this is not the case even though the documentation led me to expect it: https://github.com/curl/curl/blob/master/docs/INTERNALS.md#curl_handler

Comment (emilengler) [reply lib/gopher.c:115]: > I'm new to the curl codebase, and there's bound to be mistakes.

We're all at this point ;)

I mean why not move the function that sets the variable directly to the variable creation?

**Comment by sebdeckers** [comment]
TIMTOWTDI 🤷🏼‍♂️ Been a few months since I looked into that discussion. Has anyone begun the IETF legwork to discuss/propose a Gopher with TLS spec? I've seen a bunch of approaches with pros/cons.

**Comment by emilengler** [comment]
I would prefer to not do the discussion here, but as far as I know there hasn't been any progress made mostly due to backwards compatibility issues

**Comment by parazyd** [comment]
> I would prefer to not do the discussion here, but as far as I know there hasn't been any progress made mostly due to backwards compatibility issues

There are no compatibility issues with my patch and proposal. Daemons like geomyidae already implement this specific proposal.

**Comment by bagder** [comment]
> Sorry, I disagree. So far there is no specification for it.
> In March this year I brought up a discussion at the Gopher-ML about a specification (https://lists.debian.org/gopher-project/2020/03/msg00005.html)
> As long as there isn't anything to agree on this shouldn't be merged in my opinion.

I disagree. For me, even without a URL spec - if there is an *adopted* URL syntax already with both consumers and providers, that fact puts the code and suggestion in different angle.

@parazyd do you perhaps even already have this TLS version of the protocol documented somewhere for reference?

**Comment by parazyd** [comment]
@bagder Mostly documented in code. Here are a few links:

* https://lists.debian.org/gopher-project/2020/06/msg00001.html
* https://git.codemadness.org/hurl/commit/9546c0f17665658befbc25876245acaa9db4b08f.html
* http://r-36.net/scm/geomyidae/commit/07240d76fd8e1d0a67c49bf7e123bb508613e691.html
* http://r-36.net/scm/geomyidae/commit/89d4f0d0cf90543362a036133e88cb9fe21513d4.html

**Comment by bagder** [comment]
I volunteer to extend the test suite for gophers.

**Comment by parazyd** [comment]
@bagder If it makes sense, could you push to this branch and then we can make it a more complete merge request?

**Comment by bagder** [comment]
Yes, once I have something to show!

**Comment by bagder** [comment]
One commit that adds support for 'gophers' to the test suite. The second one provides an initial test case: that's test 1200 just updated to do gophers instead.

**Comment by bagder** [comment]
related: you should consider registering the gophers URI scheme at https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml

**Comment by bagder** [comment]
coming back to this: now we have a `docs/URL-SYNTAX.md` document that should be extended to also mention this scheme.

Anything else that needs to be done before we merge?

**Comment by parazyd** [comment]
I think we're all set if that is added :)

**Comment by bagder** [comment]
Thanks!

**Pull request closed without merging**